### PR TITLE
StudyProgramme as precondition

### PR DESCRIPTION
**Pull request by nhaagen**
Implementing feature https://www.ilias.de/docu/goto_docu_wiki_wpage_4528_1357.html

Click [here](https://github.com/ILIAS-eLearning/ILIAS/pull/568/files?w=1) to view changes without mere whitespace changes.

### CODE REVIEW

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php [hunk 1]
```diff
@@ -87,6 +90,51 @@ static function _checkGoto($a_target)
 
 		return false;
 	}
+
+	/**
+	 * Get operators
+	 */
+	public static function getConditionOperators()
+	{
+		include_once './Services/AccessControl/classes/class.ilConditionHandler.php';
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php]: Please remove this. It's included in the [classmap](https://github.com/ILIAS-eLearning/ILIAS/pull/541).

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php [hunk 2]
```diff
@@ -87,6 +90,51 @@ static function _checkGoto($a_target)
...
+	 * @param type $a_obj_id
+	 * @param type $a_operator
+	 * @param type $a_value
+	 * @param type $a_usr_id
+	 * @return boolean
+	 */
+	public static function checkCondition($a_obj_id,$a_operator,$a_value,$a_usr_id)
+	{
+		include_once './Services/AccessControl/classes/class.ilConditionHandler.php';
+		include_once "./Modules/StudyProgramme/classes/class.ilStudyProgrammeUserProgress.php";
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php]: Please remove this. They are included in the [classmap](https://github.com/ILIAS-eLearning/ILIAS/pull/541).

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php [hunk 3]
```diff
@@ -87,6 +90,51 @@ static function _checkGoto($a_target)
...
+
+		switch($a_operator) {
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php]: Please turn switch to `if ($a_operator == ...`. This is what the switch is doing effectively.

**Comment by klees** [comment]
Hi @smeyer-ilias, please have a short look into this, as it does some minor formatting in your code and also adds a new constant. Regards!

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php [hunk 4]
```diff
@@ -87,6 +90,51 @@ static function _checkGoto($a_target)
...
+					ilStudyProgrammeProgress::STATUS_COMPLETED,
+					ilStudyProgrammeProgress::STATUS_ACCREDITED
+				);
+
+				$prg_user_progress = ilStudyProgrammeUserProgress::getInstancesForUser($a_obj_id, $a_usr_id);
+				foreach ($prg_user_progress as $progress) {
+					if( in_array($progress->getStatus(), $valid_progress)) {
+						return true;
+					}
+				}
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php]: Let's take the time for a little refactoring in the SP here:

* Move the static-functions on `ilStudyProgrammeUserProgress` that query from the database to a new non static `ilStudyProgrammeUserProgressDB`.
* Look for usage sites of these static functions. Replace them by a `$this->getStudyProgrammeUserProgressDB()->...`.
* If the object using the static function is one where we control the creation ourselves, pass the new DB as a dependency in the constructor and implement `getStudyProgrammeUserProgressDB` accordingly. If not, implement `getStudyProgrammeUserProgressDB` by creating and caching an instance. If a static function uses the statics, implement `getStudyProgrammeUserProgressDB` as static as well and call it as `static::getStudyProgrammeUserProgressDB`. This will allow to overwrite the method for a test.
* This should open some opportunities for tests. Write them.
* This will make it possible to test this very method. Write the test.

**Review by klees — COMMENTED** [review]
LGTM in general, let's use the refactoring opportunity.

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php (hunk 4 above)
Comment (smeyer-ilias) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php]: A note, about performance: this method is called for every 'visible' and 'read' permission check of SPs with preconditions. Maybe a caching would improve performance. Additionally one could use ilObjectAccess::_preloadData($obj_ids, $ref_ids) to load required data (for _checkCondition, _checkAccess) from the database in one step.

File: Services/AccessControl/classes/class.ilConditionHandlerTableGUI.php [hunk 5]
```diff
@@ -122,19 +122,18 @@ protected function initTable()
...
 		{
 			$this->addCommandButton("saveObligatoryList", $this->lng->txt("rbac_precondition_save_obligatory"));
 		}
 	}
 
-	
 }
-?>
\ No newline at end of file
+?>
```
Comment (smeyer-ilias) [inline Services/AccessControl/classes/class.ilConditionHandlerTableGUI.php]: Please revert this file (only whitespace changes).

File: Services/AccessControl/classes/class.ilConditionSelector.php [hunk 6]
```diff
@@ -124,7 +124,4 @@ function isNodeHighlighted($a_node)
 		return parent::isNodeHighlighted($a_node);
 	}
 
-
-
-
-} 
\ No newline at end of file
+}
```
Comment (smeyer-ilias) [inline Services/AccessControl/classes/class.ilConditionSelector.php]: Please revert this file (only whitespace changes).

**Review by smeyer-ilias — CHANGES_REQUESTED** [review]
Thanks, only small whitespace changes and one note regarding performance.

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php (hunk 4 above)
Comment (klees) [reply Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php]: Hi @nhaagen, lets have a look into this together. No hurry, though.

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php [hunk 7]
```diff
@@ -87,6 +107,44 @@ static function _checkGoto($a_target)
...
+	 */
+	public static function checkCondition($a_obj_id,$a_operator,$a_value,$a_usr_id)
+	{
+		if ($a_operator === ilConditionHandler::OPERATOR_ACCREDITED_OR_PASSED) {
+			$valid_progress = array(
+				ilStudyProgrammeProgress::STATUS_COMPLETED,
+				ilStudyProgrammeProgress::STATUS_ACCREDITED
+			);
+
+			$prg_user_progress = $this->getStudyProgrammeUserProgressDB()->getInstancesForUser($a_obj_id, $a_usr_id);
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php]: There is no `$this` in static functions. Or is this called non-statically?

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php [hunk 8]
```diff
@@ -14,7 +16,25 @@
  * @author: Richard Klees <richard.klees@concepts-and-training.de>
  *
  */
-class ilObjStudyProgrammeAccess extends ilObjectAccess {
+class ilObjStudyProgrammeAccess extends ilObjectAccess implements ilConditionHandling{
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php]: One spaces before the `{` please.

File: Modules/StudyProgramme/classes/class.ilObjStudyProgramme.php [hunk 9]
```diff
@@ -55,36 +60,63 @@ public function __construct($a_id = 0, $a_call_by_reference = true) {
...
+		return $this->sp_user_progress_db;
+	}
+
+	/**
+	* Get a new instance of ilStudyProgrammeUserProgressDB
+	*
+	* @return ilStudyProgrammeUserProgressDB
+	*/
+	static public function getNewStudyProgrammeUserProgressDB() {
+		require_once("./Modules/StudyProgramme/classes/class.ilStudyProgrammeUserProgressDB.php");
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgramme.php]: Not required anymore, we use autoloading with classmap now.

File: Modules/StudyProgramme/classes/class.ilObjStudyProgramme.php [hunk 10]
```diff
@@ -55,36 +60,63 @@ public function __construct($a_id = 0, $a_call_by_reference = true) {
...
+	*/
+	public function getStudyProgrammeUserProgressDB() {
+		if(! $this->sp_user_progress_db) {
+			$this->sp_user_progress_db = static::getNewStudyProgrammeUserProgressDB();
+		}
+		return $this->sp_user_progress_db;
+	}
+
+	/**
+	* Get a new instance of ilStudyProgrammeUserProgressDB
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgramme.php]: This is lying. Only one fresh instance is created.

File: Modules/StudyProgramme/classes/class.ilObjStudyProgramme.php [hunk 11]
```diff
@@ -55,36 +60,63 @@ public function __construct($a_id = 0, $a_call_by_reference = true) {
...
+	*
+	* @return ilStudyProgrammeUserProgressDB
+	*/
+	static public function getNewStudyProgrammeUserProgressDB() {
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgramme.php]: The name also contains the missleading `New`. To avoid a nameclash with the non-static method we could use an (ugly) "_" in front of the name...

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php [hunk 12]
```diff
@@ -14,7 +16,25 @@
...
+	 * @var ilStudyProgrammeUserProgressDB
+	 */
+	protected $sp_user_progress_db;
+
+	/**
+	* Get a (cached) instance of ilStudyProgrammeUserProgressDB
+	*
+	* @return ilStudyProgrammeUserProgressDB
+	*/
+	private function getStudyProgrammeUserProgressDB() {
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php]: This needs to become static, because we should not call it from `checkCondition` otherwise. Also you do not need to cache the DB, `ilObjStudyProgramme` already does that. I just realised that this also means that this complete method could just go away an be replaced by the according method on `ilObjStudyProgramme`.

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php [hunk 13]
```diff
@@ -87,6 +107,44 @@ static function _checkGoto($a_target)
...
+	public static function getConditionOperators()
+	{
+		return array(
+			ilConditionHandler::OPERATOR_ACCREDITED_OR_PASSED
+		);
+	}
+
+	/**
+	 *
+	 * @global ilObjUser $ilUser
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeAccess.php]: ?

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeIndividualPlanGUI.php [hunk 14]
```diff
@@ -83,21 +88,33 @@ public function __construct($a_parent_gui, $a_ref_id) {
...
+	* Get a (cached) instance of ilStudyProgrammeUserProgressDB
+	*
+	* @return ilStudyProgrammeUserProgressDB
+	*/
+	private function getStudyProgrammeUserProgressDB() {
+		if(! $this->sp_user_progress_db) {
+			 $this->sp_user_progress_db = ilObjStudyProgramme::getNewStudyProgrammeUserProgressDB();
+		}
+		return $this->sp_user_progress_db;
+	}
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeIndividualPlanGUI.php]: We control the creation of `ilObjStudyProgrammeIndividualPlabGUI` so we could inject this dependency. I see no additional value in this getter, then, so you could use `$this->sp_user_progress_db` directly then.

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeMembersGUI.php [hunk 15]
```diff
@@ -87,29 +92,41 @@ public function __construct($a_parent_gui, $a_ref_id) {
...
 		$lng->loadLanguageModule("prg");
 	}
-	
+
+	/**
+	* Get a (cached) instance of ilStudyProgrammeUserProgressDB
+	*
+	* @return ilStudyProgrammeUserProgressDB
+	*/
+	public function getStudyProgrammeUserProgressDB() {
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeMembersGUI.php]: Why is this public here?

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeMembersGUI.php [hunk 16]
```diff
@@ -87,29 +92,41 @@ public function __construct($a_parent_gui, $a_ref_id) {
...
+		if(! $this->sp_user_progress_db) {
+			 $this->sp_user_progress_db = ilObjStudyProgramme::getNewStudyProgrammeUserProgressDB();
+		}
+		return $this->sp_user_progress_db;
+	}
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeMembersGUI.php]: See comment at according method in `ilObjStudyProgrammeIndividualPlanGUI`.

File: Modules/StudyProgramme/classes/class.ilObjStudyProgrammeMembersGUI.php [hunk 17]
```diff
@@ -302,43 +319,43 @@ protected function initSearchGUI() {
...
 	 * @param	int		$a_ass_id		Id of the assignment object to act on.
 	 * @return	string					The link to the action.
 	 */
 	public function getLinkTargetForAction($a_action, $a_prgrs_id, $a_ass_id) {
 		require_once("Modules/StudyProgramme/classes/class.ilStudyProgrammeUserProgress.php");
-		
+
 		switch ($a_action) {
-			case ilStudyProgrammeUserProgress::ACTION_MARK_ACCREDITED:
+			case ilStudyProgrammeUserProgressDB::ACTION_MARK_ACCREDITED:
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilObjStudyProgrammeMembersGUI.php]: Why did you move these to the `DB`-class?

File: Modules/StudyProgramme/classes/class.ilStudyProgrammeUserProgressDB.php [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+			$actions[] = self::ACTION_REMOVE_USER;
+		}
+		if ($a_status == ilStudyProgrammeProgress::STATUS_ACCREDITED) {
+			$actions[] = self::ACTION_UNMARK_ACCREDITED;
+		}
+		else if ($a_status == ilStudyProgrammeProgress::STATUS_IN_PROGRESS) {
+			$actions[] = self::ACTION_MARK_ACCREDITED;
+		}
+		return $actions;
+	}
```
Comment (klees) [inline Modules/StudyProgramme/classes/class.ilStudyProgrammeUserProgressDB.php]: I know I told you to move all statics to this new class, but this most probably does not belong to a `DB`-class. Do we ever need that without a concrete instance of `ilStudyProgrammeUserProgress`? If not, this could as well become a non-static method on the class.

File: Modules/StudyProgramme/test/ilStudyProgrammeUserProgressTest.php [hunk 19]
```diff
@@ -593,25 +593,110 @@ public function testCompletionOnDeeplyNestedProgresses() {
...
+		);
+
+		$up = $this->root->getProgressesOf($user->getId())[0];
+		$this->assertEquals(
+			$up,
+			$sp_user_progress_db->getInstanceById($up->getId())
+		);
+	}
+
+	public function testGetInstanceCalls() {
```
Comment (klees) [inline Modules/StudyProgramme/test/ilStudyProgrammeUserProgressTest.php:688]: These are two tests.

File: Services/AccessControl/classes/class.ilConditionHandler.php [hunk 20]
```diff
@@ -433,24 +434,24 @@ function enableAutomaticValidation($a_validate = true)
...
 		foreach($objDefinition->getPlugins() as $p_type => $p_info)
 		{
 			if(@include_once $p_info['location'].'/class.ilObj'.$p_info['class_name'].'Access.php')
 			{
 				include_once './Services/AccessControl/interfaces/interface.ilConditionHandling.php';
 				$name = 'ilObj'.$p_info['class_name'].'Access';
-				$refection = new ReflectionClass($name);
-				if($refection->implementsInterface('ilConditionHandling'))
+				$reflection = new ReflectionClass($name);
+				if($reflection->implementsInterface('ilConditionHandling'))
```
Comment (klees) [inline Services/AccessControl/classes/class.ilConditionHandler.php:446]: 👍  for removing spelling mistakes.

Please revert the whitespace changes, as @smeyer-ilias asked us.

File: Services/AccessControl/classes/class.ilConditionHandlerGUI.php [hunk 21]
```diff
@@ -882,8 +886,8 @@ private function initFormCondition($a_source_id,$a_condition_id = 0,$a_mode = 'a
 	 			$this->form->addCommandButton('updateCondition',$this->lng->txt('save'));
 	 			$this->form->addCommandButton('listConditions',$this->lng->txt('cancel'));
 	 			break;
-	 			
-	 		
+
+
```
Comment (klees) [inline Services/AccessControl/classes/class.ilConditionHandlerGUI.php]: Please revert whitespace changes as @smeyer-ilias asked us.

File: lang/ilias_de.lang [hunk 22]
```diff
@@ -41,7 +41,7 @@ common#:#pd_enable_pdfrmpostdraft_info#:#Sofern aktiviert, wird auf dem Persönl
 pd#:#block_show_pdfrmpostdraft#:#'Meine Beitragsentwürfe' anzeigen
 forum#:#forums_thread_drafts_deleted#:#Das gewählte Thema einschließlich aller dazugehörigen Beiträge und evtl. Beitragsentwürfe wurde gelöscht.
 forum#:#forums_threads_drafts_deleted#:#Die gewählten Themen einschließlich aller dazugehörigen Beiträge und evtl. Beitragsentwürfe wurden gelöscht.
-forum#:#save_draft_successfully#:#Der Entwurf wurde erfolgreich gespeichert. 
+forum#:#save_draft_successfully#:#Der Entwurf wurde erfolgreich gespeichert.
```
Comment (klees) [inline lang/ilias_de.lang]: We seem to need a less aggressive whitespace purging, please remove the whitespace changes in this file as well.

Comment (klees) [reply lang/ilias_de.lang]: Is there any real change in this file?

**Review by klees — COMMENTED** [review]
Thx for refactoring. LGTM in general. Some stuff needs to be done regarding the refactoring. Please remove the whitespace changes in files that we do not maintain.
